internal/controller/metalstackmachine_controller.go

```diff
@@ -392,6 +400,40 @@ func (r *machineReconciler) findProviderMachine() (*models.V1MachineResponse, er
 	}
 }
 
+func (r *machineReconciler) patchMachineLabels(m *models.V1MachineResponse) {
+	if r.infraMachine.Labels == nil {
+		r.infraMachine.Labels = make(map[string]string)
+	}
+
+	if m.Allocation != nil && m.Allocation.Hostname != nil {
+		r.infraMachine.Labels[corev1.LabelHostname] = *m.Allocation.Hostname
+	}
+	if m.Partition != nil && m.Partition.ID != nil {
+		r.infraMachine.Labels[corev1.LabelTopologyZone] = *m.Partition.ID
+	}
+
+	if m.Partition != nil && m.Partition.Labels != nil && m.Partition.Labels[tag.PartitionRegion] != "" {
+		r.infraMachine.Labels[corev1.LabelTopologyRegion] = m.Partition.Labels[tag.PartitionRegion]
+	}
+
+	tagMap := tag.NewTagMap(m.Tags)
+
+	rack := m.Rackid
+	if rack == "" {
+		rack, _ = tagMap.Value(tag.MachineRack)
+	}
+	if rack != "" {
+		r.infraMachine.Labels[tag.MachineRack] = rack
+	}
```

Comment:
I don't understand why we have two source for the rack.

Reply:
Found this confusing, too. And with pointers in APIs I usually go defensively.
 I'd go with `Rackid`.

Reply:
Tags of a machine are automatically synced to the node object by the metal-ccm, therefore we did add some stuff to the machine tags in the past.